src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,545 @@
+                return editor.GetChangedRoot();
+            };
+        }
+
+        private static PatternSyntax RewriteContainingPattern(
+            PatternSyntax containingPattern,
+            PatternSyntax generatedPattern,
+            ImmutableArray<IdentifierNameSyntax> namesOpt)
+        {
+            // This is a variable designation match. We'll try to combine the generated
+            // pattern from the right-hand-side into the containing pattern of this designation.
+            var rewrittenPattern = namesOpt.IsDefault
+                // If there's no name, we will combine the pattern itself.
+                ? Combine(containingPattern, generatedPattern)
+                // Otherwise, we generate a subpattern per each name and rewrite as a recursive pattern.
+                : AddSubpattern(containingPattern, CreateSubpattern(namesOpt, generatedPattern));
+
+            // We must have preserved the existing variable designation.
+            Debug.Assert(containingPattern switch
+            {
+                VarPatternSyntax p => p.Designation,
+                DeclarationPatternSyntax p => p.Designation,
+                RecursivePatternSyntax p => p.Designation,
+                var p => throw ExceptionUtilities.UnexpectedValue(p)
+            } is var d && rewrittenPattern.DescendantNodes().Any(node => AreEquivalent(node, d)));
+
+            return rewrittenPattern
+                .WithAdditionalAnnotations(Formatter.Annotation)
+                .WithAdditionalAnnotations(Simplifier.Annotation);
+
+            static PatternSyntax Combine(PatternSyntax containingPattern, PatternSyntax generatedPattern)
+            {
+                return (containingPattern, generatedPattern) switch
+                {
+                    // We know we have a var-pattern, declaration-pattern or a recursive-pattern on the left as the containing node of the variable designation.
+                    // Depending on the generated pattern off of the expression on the right, we can give a better result by morphing it into the existing match.
+
+                    // e.g. `e is var x && x is { p: 1 }` => `e is { p: 1 } x`
+                    (VarPatternSyntax var, RecursivePatternSyntax { Designation: null } recursive)
+                        => recursive.WithDesignation(var.Designation),
+
+                    // e.g. `e is var x && x is C` => `e is C x`
+                    (VarPatternSyntax var, TypePatternSyntax type)
+                        => DeclarationPattern(type.Type, var.Designation),
+
+                    // e.g. `e is C x && x is { p: 1 }` => `is C { p: 1 } x`
+                    (DeclarationPatternSyntax decl, RecursivePatternSyntax { Type: null, Designation: null } recursive)
+                        => recursive.WithType(decl.Type).WithDesignation(decl.Designation),
+
+                    // e.g. `e is { p: 1 } x && x is C` => `is C { p: 1 } x`
```

Comment:
coudl this cahnge meaning though if `C` is a different type than `e`s type?  One is checking the props of the static e type, the other is checking after casting to C and checking it's properties.